AK/Math.h

```diff
@@ -424,48 +432,70 @@ using Trigonometry::tan;
 namespace Exponentials {
 
 template<FloatingPoint T>
-constexpr T log(T x)
+constexpr T log2(T x)
 {
-    CONSTEXPR_STATE(log, x);
+    CONSTEXPR_STATE(log2, x);
 
 #if ARCH(X86_64)
     T ret;
     asm(
-        "fldln2\n"
+        "fld1\n"
         "fxch %%st(1)\n"
         "fyl2x\n"
         : "=t"(ret)
         : "0"(x));
     return ret;
 #else
 #    if defined(AK_OS_SERENITY)
-    // TODO: Add implementation for this function.
-    TODO();
+    if (x == 0)
+        return Infinity<T>;
+    if (x <= 0 || __builtin_isnan(x))
+        return NaN<T>;
+
+    // FIXME: Make this more accurate, currently it shows errors of up to 0.0017
+    //        If this is done, use it for x86 floats/doubles as well
+    FloatExtractor<T> ext { .d = x };
+
+    FloatExtractor<T> mantissa_ext {
+        .sign = ext.sign,
+        .exponent = FloatExtractor<T>::exponent_bias,
+        .mantissa = ext.mantissa
+    };
+    // (1 <= mantissa < 2)
+    T mantissa = mantissa_ext.d;
+
+    // Approximation using the 4th degree taylor approximation around 1.5
+    T x1 = mantissa - 1.5;
+    T x2 = x * x;
+    T x3 = x2 * x;
+    T x4 = x3 * x;
+    T log2_mantissa = (T)0.584963 + (T)0.961797 * x1 - (T)0.320599 * x2 + (T)0.142488 * x3 - (T)0.0712442 * x4;
```

Comment:
In similar contexts, [Horner's scheme](https://en.wikipedia.org/wiki/Horner%27s_method) to evaluate polynomials often results in better behavior, especially if weird rounding happens. Did you also measure that behavior?

```suggestion
T log2_mantissa = ((((T)0.0712442 * x1 + (T)0.142488) * x1 - (T)0.320599) * x1 + (T)0.961797) * x1 + (T)0.584963;
```

Also, some comments about how you derived these constants would be nice.
Also, I'm surprised be the low amount of significant digits. I would have guessed that just writing more digits would make it slightly more accurate?

Reply:
This doesn't look like a good approximation at all: https://www.wolframalpha.com/input?i=plot+0.584963+%2B+0.961797+*+%28x-1.5%29+-+0.320599+*+x%5E2+%2B+0.142488+*+x%5E3+-+0.0712442+*+x%5E4+-+log2%28x%29+for+x+from+1+to+2

If I evaluate the derivatives of log2(x), then I get the constants:
```
f(x) = log_2(x) = ln(x) / ln(2)
    f(1.5) / 0! = 0.5849625007211562
f'(x) = 1/(ln(2)*x)
    f'(1.5) / 1! = 0.9617966939259757
f''(x) = -1/(ln(2)*x^2)
    f''(1.5) / 2! = -0.32059889797532526
f'''(x) = 1/(ln(2)*x^3)
    f'''(1.5) / 3! = 0.07124419955007229
f''''(x) = -1/(ln(2)*x^4)
    f''''(1.5) / 4! = -0.011874033258345379
```

These are suspiciously similar to the constants you used. Is it possible that there was an off-by-one-error in your computations?

Using these values and the correct coefficients, it looks like a much better fit: https://www.wolframalpha.com/input?i=plot+0.5849625007211562+%2B+0.9617966939259757+*+%28x-1.5%29+-+0.32059889797532526+*+%28x-1.5%29%5E2+%2B+0.07124419955007229+*+%28x-1.5%29%5E3+-+0.011874033258345379+*+%28x-1.5%29%5E4+-+log2%28x%29+for+x+from+1+to+2

Note that this results in an absolute error of [-0.0061, 0.014]. (Relative error isn't defined because log2(1) = 0.)

Weirdly, my guess to use 1.414… as expansion point leads to a significantly worse result:
```
f(x) = log_2(x) = ln(x) / ln(2)
    f(sqrt(2)) / 0! = 0.5   (by definition!)
f'(x) = 1/(ln(2)*x)
    f'(sqrt(2)) / 1! = 1.0201394465967895
f''(x) = -1/(ln(2)*x^2)
    f''(sqrt(2)) / 2! = -0.7213475204444816
f'''(x) = 1/(ln(2)*x^3)
    f'''(sqrt(2)) / 3! = 0.5100697232983946
f''''(x) = -1/(ln(2)*x^4)
    f''''(sqrt(2)) / 4! = -0.36067376022224074
```

(Absolute error about -0.09.)
https://www.wolframalpha.com/input?i=plot+0.5+%2B+1.0201394465967895+*+%28x-sqrt%282%29%29+-+0.7213475204444816+*+%28x-sqrt%282%29%29%5E2+%2B+0.5100697232983946+*+%28x-sqrt%282%29%29%5E3+-+0.36067376022224074+*+%28x-sqrt%282%29%29%5E4+-+log2%28x%29+for+x+from+1+to+2

Reply:
Come to think of it, why a Taylor expansion at all? That would yield great accuracy around the expansion point, yes, but unless we manually tweak the coefficients it would also create discontinuous jumps around the wraparound, so log2(16-epsilon) and log2(16+epsilon) are far apart, no matter how small epsilon becomes.

How about using a [cubic polynomial](https://en.wikipedia.org/wiki/Cubic_Hermite_spline) instead? It's cheaper to compute, and not only is it continuous, it's even continuous in its derivative!

```
// (3/(2*ln(2)) - 2) x³ + (-7/ln(2)+9) x² + (21/(2*ln(2)) - 12)x  - 5/ln(2)+5
T log_2_mantissa = ((0.1640425613334453 * x - 1.098865286222745) * x + 3.1482979293341167) * x - 2.2134752044448174;
```

https://www.wolframalpha.com/input?i=plot+%283%2F%282*ln%282%29%29+-+2%29+x%C2%B3+%2B+%28-7%2Fln%282%29%2B9%29+x%C2%B2+%2B+%2821%2F%282*ln%282%29%29+-+12%29x++-+5%2Fln%282%29%2B5+-log2%28x%29+for+x+from+1+to+2

It's even got a better accuracy, the error is only 0.005!